<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-необходимых-библиотек" data-toc-modified-id="Импорт-необходимых-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт необходимых библиотек</a></span></li><li><span><a href="#Открытие-файлов" data-toc-modified-id="Открытие-файлов-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Открытие файлов</a></span></li><li><span><a href="#Вывод-десяти-строк-таблиц" data-toc-modified-id="Вывод-десяти-строк-таблиц-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод десяти строк таблиц</a></span></li><li><span><a href="#Вывод-основной-информации-о-датафрейме" data-toc-modified-id="Вывод-основной-информации-о-датафрейме-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод основной информации о датафрейме</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Разбиение-данных-на-целевой-признак-и-признаки" data-toc-modified-id="Разбиение-данных-на-целевой-признак-и-признаки-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Разбиение данных на целевой признак и признаки</a></span></li><li><span><a href="#Преобразование-данных" data-toc-modified-id="Преобразование-данных-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Преобразование данных</a></span></li><li><span><a href="#Разбиение-данных-на-выборки" data-toc-modified-id="Разбиение-данных-на-выборки-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Разбиение данных на выборки</a></span></li><li><span><a href="#Обучение-и-проверка-моделей" data-toc-modified-id="Обучение-и-проверка-моделей-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Обучение и проверка моделей</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Финальные-комментарии-ревьюера" data-toc-modified-id="Финальные-комментарии-ревьюера-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Финальные комментарии ревьюера</a></span></li></ul></div>

# Защита персональных данных клиентов

Нужно защитить данные клиентов страховой компании «Хоть потоп». Необходимо разработать такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию, а также обосновать корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

**Цель исследования** — найти такой метод преобразования данных, при котором данные было бы сложно восстановить, а качество моделей машинного обучения не ухудшилось бы.

**Ход исследования:**
1. Загрузка данных
2. Умножение матриц
3. Алгоритм преобразования
4. Проверка алгоритма
5. Вывод

## Загрузка данных

### Импорт необходимых библиотек

In [66]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split 
from sklearn.metrics import r2_score

### Открытие файлов

In [67]:
try:
    data = pd.read_csv('/datasets/insurance.csv', engine='python')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/insurance.csv', engine='python')

### Вывод десяти строк таблиц

In [68]:
data.head(10)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
5    1     43.0   41000.0            2                  1
6    1     39.0   39700.0            2                  0
7    1     25.0   38600.0            4                  0
8    1     36.0   49700.0            1                  0
9    1     32.0   51700.0            1                  0

### Вывод основной информации о датафрейме

In [69]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [70]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [71]:
data.duplicated().sum()

153

### Вывод

Согласно документации к данным, в датасете содержатся такие признаки, как пол, возраст и зарплата застрахованного, количество членов его семьи, а также целевой признак — количество страховых выплат клиенту за последние 5 лет.
В данных есть 153 дублирующиеся строки, но они не требуют предобработки, так как данные в данном случае просто могут повторяться.
Данные в полном порядке и не требуют предобработки.

## Умножение матриц

Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** Качество линейной регрессии не изменится.<a id='par2'></a>

**Обоснование:** Чтобы узнать, изменится ли качество линейной регресии, необходимо сравнить предсказания до и после умножения признаков на обратимую матрицу.

Предсказания до умножения признаков на обратимую матрицу:

$$
a_1 = Xw_1,
$$

где w<sub>1</sub> — вектор весов линейной регрессии до умножения признаков на обратимую матрицу:

$$
w_1 = (X^T X)^{-1} X^T y
$$

Предсказания после умножения признаков на обратимую матрицу:

$$
a_2 = Cw_2,
$$

где w<sub>2</sub> — вектор весов линейной регрессии после умножения признаков на обратимую матрицу:

$$
w_2 = (C^T C)^{-1} C^T y,
$$

а C — это произведение признаков и обратимой матрицы:

$$
C = X P
$$

Раскроем скобки в формуле предсказаний a<sub>2</sub> с помощью формулы произведения признаков и обратимой матрицы и формулы вектора весов линейной регресии:

$$
a_2 = X P ((X P)^T X P)^{-1} (X P)^T y
$$

Затем раскроем скобки по свойству транспонирования матриц:

$$
a_2 = X P (P^T X^T X P)^{-1} P^T X^T y
$$

Так как X<sup>T</sup>X, P<sup>T</sup> и P — это квадратные матрицы, то по свойству обратных матриц раскроем скобки:

$$
a_2 = X P P^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y
$$

(P<sup>T</sup>)<sup>-1</sup> P<sup>T</sup> = E, поэтому (P<sup>T</sup>)<sup>-1</sup> и P<sup>T</sup> можно убрать:

$$
a_2 = X P P^{-1} (X^T X)^{-1} X^T y
$$

P P<sup>-1</sup> = E, P и P<sup>-1</sup> тоже убираем:

$$
a_2 = X (X^T X)^{-1} X^T y
$$

$$
a_1 = Xw_1 = X (X^T X)^{-1} X^T y = a_2
$$

Таким образом, предсказания до и после умножения признаков на обратимую матрицу равны, следовательно, качество линейной регресии не изменится.

## Алгоритм преобразования

**Алгоритм**

Исходя из пункта [2](#par2), можно преобразовать данные с помощью умножения признаков на обратимую матрицу. При преобразовании качество моделей машинного обучения не ухудшится, а также данные можно будет восстановить после преобразования, умножив их на матрицу, обратную изначальной обратимой матрице.

1. Сгенерировать случайную квадратную матрицу, размер которой будет соответствовать ширине матрицы признаков.
2. Проверить сгенерированную матрицу на обратимость.
3. Умножить матрицу признаков на обратимую матрицу.
4. Обучить одну модель на стандартных данных и другую — на преобразованных данных, затем получить предсказания.
5. Проверить обе модели с помощью метрики R2.
5. Сравнить получившиеся метрики R2 до и после умножения признаков на обратимую матрицу.

## Проверка алгоритма

### Разбиение данных на целевой признак и признаки

In [72]:
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']

### Преобразование данных

Сгенерируем квадратную обратимую матрицу и проверим ее обратимость:

In [73]:
while True:
    invert_matrix = np.random.normal(size=(features.shape[1], features.shape[1]))
    try:
        np.linalg.inv(invert_matrix)
        print('Кодирующая матрица:')
        print(invert_matrix)
        break
    except:
        continue

Кодирующая матрица:
[[-0.00427926  0.37830843  1.45394668 -0.31026098]
 [ 0.76874006 -0.85134667  0.17477458 -0.71698321]
 [ 0.19438479 -1.62239111 -0.2967114   0.83608592]
 [ 0.91717961 -0.17769812  1.5862778   0.67427151]]


Преобразуем признаки с помощью умножения на обратимую матрицу:

In [74]:
new_features = features @ invert_matrix

### Разбиение данных на выборки

Разобьем стандартные и преобразованные данные для двух моделей соответственно в отношении 75:25 на обучающую и валидационную выборки:

In [75]:
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=12345)

new_features_train, new_features_valid, new_target_train, new_target_valid = train_test_split(
    new_features, target, test_size=0.25, random_state=12345)

In [76]:
print(features_train.shape, features_valid.shape)
print(new_features_train.shape, new_features_valid.shape)

(3750, 4) (1250, 4)
(3750, 4) (1250, 4)


### Обучение и проверка моделей

Обучим модель на еще не преобразованных данных, получим предсказания и проверим модель с помощью метрики R2:

In [77]:
model = LinearRegression()
model.fit(features_train, target_train)
y_pred = model.predict(features_valid)
print(r2_score(target_valid, y_pred))

0.43522757127026546


Теперь сделаем то же самое для преобразованных данных:

In [78]:
model.fit(new_features_train, new_target_train)
new_y_pred = model.predict(new_features_valid)
print(r2_score(new_target_valid, new_y_pred))

0.4352275712699596


### Вывод

Проверка моделей линейной регрессии с помощью метрики r2 подтвердила вывод из пункта [2](#par2) — модель, обученная и проверенная на изначальных данных, и модель, обученная и проверенная на преобразованных данных, дали одинаковый результат.

## Вывод

С помощью математических формул было доказано, что при умножении признаков на обратимую матрицу качество моделей машинного обучения не ухудшилось, поэтому алгоритмом преобразования данных было выбрано произведение признаков и обратимой матрицы.

После преобразования данных и разбиения их на выборки были обучены две модели линейной регресии: одна модель обучалась и проверялась на стандартных данных, а другая — на преобразованных данных. Затем модели были проверены с помощью метрики r2 — для обеих моделей метрика оказалась одинаковой, следовательно качество модели от преобразования не ухудшилось.